navigation-menu: apply colors correctly in edition mode

retrofox opened the pull request:
## Description
This PR ~fixes~ implements the following changes:

* Align the `Aa` text of the toolbar colors-selector button.
* Apply the styles via CSS classes when the colores comes from the color palette.
* Apply the styles via inline styles when the colors are custom one.

These changes are implemented in the Editor Canvas. Let's handle the front-end in another PR.

## Screenshots 

**before**
![image](https://user-images.githubusercontent.com/77539/67900555-838dce00-fb43-11e9-96cf-84ee83cd7781.png)

![image](https://user-images.githubusercontent.com/77539/67900996-6d344200-fb44-11e9-9721-01e340f4e1a0.png)


**after**
![image](https://user-images.githubusercontent.com/77539/67900611-a3bd8d00-fb43-11e9-90d9-fc39567508fa.png)

![image](https://user-images.githubusercontent.com/77539/67901063-9523a580-fb44-11e9-9472-2d521e95da64.png)



## Checklist:
- [x] My code is tested.
- [x] My code follows the WordPress code style. <!-- Check code: `npm run lint`, Guidelines: https://make.wordpress.org/core/handbook/best-practices/coding-standards/javascript/ -->

retrofox committed:
navigation-menu: apply colors correctly in edition mode

retrofox committed:
navigation-menu: remove unneeded css width rule

talldan commented on packages/block-library/src/navigation-menu/block-colors-selector.js:
```diff
@@ -20,7 +20,7 @@ import { ColorPaletteControl, ContrastChecker } from '@wordpress/block-editor';
 	<div className="block-library-colors-selector__icon-container">
 		<div
-			className="block-library-colors-selector__state-selection wp-block-navigation-menu-item"
+			className="block-library-colors-selector__state-selection"
```
I think it might be good to make the icon an SVG. As it is, I'm not sure there's any way to guarantee which font will be used here, and an SVG might be a better way to ensure the `Aa` is correctly centred.

retrofox replied:
can we handle this in another PR? Glad to make an issue.

retrofox replied:
Resolve this conversation in favor of https://github.com/WordPress/gutenberg/issues/18197

talldan replied:
Separate PR sounds good, thanks for making the issue. 👍

tellthemachines commented on packages/block-library/src/navigation-menu-item/editor.scss:
```diff
@@ -50,14 +50,22 @@
+		font-family: $editor-font;
+		font-weight: bold;
+		font-size: $text-editor-font-size;
+		background-color: var(--background-color-menu-link);
```
I'm afraid we can't use CSS variables because they're not supported in IE. We'll have to set these as inline styles with JS instead.

tellthemachines commented on packages/block-library/src/navigation-menu-item/editor.scss:
```diff
@@ -50,14 +50,22 @@
+	}
 
-	&:focus {
+	&.is-editing .components-text-control__input {
```
Alternatively, for the input, we could set `background-color: transparent` and `color: inherit`.

talldan replied:
I think the difference is that one way the whole block has the background color, and the other only the input does. Not sure which is better.

retrofox replied:
yes, in other words, with CSS vars we're applying the colors just to `NavigationMenuItem`, not to the `NavigationMenu`

tellthemachines reviewed (changes requested):
Thanks for addressing this bug! Left a couple comments; the major issue here is with IE support 😢

talldan commented:
@retrofox lets defer to @tellthemachines' review and replace the CSS vars with inline styles, I agree it's important we don't forget IE support. I mistakenly thought that was being fixed on https://github.com/WordPress/gutenberg/pull/18150.

marekhrabe commented:
I think that setting the background and font color only on the nav block (where it is configured using block attributes) and propagating it only through CSS to appropriate nav items would be the easiest way forward. Skipping the need to pass props between parent-child blocks (unexplored territory) and also skipping CSS vars. I think they would be best to do the job but the current support still includes IE11 so we need to make sure it works.

draganescu commented:
We could not have this feature available in IE, and move ahead implementing it using tools available in modern browsers. That means the actual button in the toolbar would not be available in IE. @shaunandrews @mtias

retrofox committed:
navigation-menu: do not use CSS vars to apply styles

retrofox committed:
naviugation-menu: apply palette colors by CSS classes

retrofox committed:
colors-selector: do not use CSS var to set colors

retrofox committed:
colors-selector: fix colors toolbar button

retrofox committed:
navigation-menu-item: remove CSS vars

retrofox committed:
navigation-menu-item: set text field colors

retrofox commented:
:wave: I've added some changes applying the styles in the navigation menu, instead of each item:

* It will, indirectly, help to make this feature available for IE. 
* It's possible that we want to apply specific styles for each item. Think for instance make the `Home` link different to the rest of others. Changes on this PR are aligned with this idea.

tellthemachines reviewed (commented):
This is now working on IE 🎉 🎉 

I'm not able to preview most of the colours on the editor side though. This happens on all browsers and I think it's due to the inline styles not being applied as per my comment below.

Another slight issue that might pop up, depending on how theme styles are set, is the background colour not being applied to the submenu dropdowns:
<img width="313" alt="Screen Shot 2019-11-04 at 4 48 17 pm" src="https://user-images.githubusercontent.com/8096000/68101965-911ccc80-ff24-11e9-98b5-d7d1a94b5017.png">
It might be worth applying the inline styles to the submenu `<ul>`s too.

Question: should the colours reset to default when switching themes? Right now it's not happening but colours don't always translate 100% from one theme to another, say if accent colour was set as background in one theme it doesn't necessarily stay set in another. I am also noticing some intermittent weirdness when setting custom colours and then switching themes. This happened with a green custom accent set in Twenty Nineteen, and then switching to Twenty Twenty:
<img width="615" alt="Screen Shot 2019-11-04 at 5 06 49 pm" src="https://user-images.githubusercontent.com/8096000/68102353-66337800-ff26-11e9-8d40-33182a1e50a4.png">

retrofox commented on packages/block-library/src/navigation-menu/edit.js:
```diff
@@ -60,19 +60,21 @@ function NavigationMenu( {
 
-	if ( backgroundColor.color ) {
-		navigationMenuStyles[ '--background-color-menu-link' ] = backgroundColor.color;
+	if ( attributes.customBackgroundColor ) {
```
> customTextColor and customBackgroundColor don't seem to be coming through at all in editor mode.
it does, when it sets custom colors:

<img src="https://user-images.githubusercontent.com/77539/68160287-35f1d680-ff32-11e9-96ae-19a20edd3115.png" width="400px" />

<img src="https://user-images.githubusercontent.com/77539/68160255-1b1f6200-ff32-11e9-9875-519ddaf876ce.png" width="400px" />

> Could we use textColor and backgroundColor instead?

`textColor` and `backgroundColor` are to apply the styles but with CSS classes instead of inline styles.

retrofox replied:
Handled here: https://github.com/WordPress/gutenberg/pull/18172/commits/91a507ed3ad408b8f5ec75409fad312015cfa2b3 🙇

retrofox commented:
@tellthemachines thanks for your comments here, and they make totally sense but I'd like to handle the front-end in a different PR. Make sense for you?
I've added the `These changes are implemented in the Editor Canvas. Let's handle the front-end in another PR.` comment in the PR description.
I've rebased the branch with master as well.

tellthemachines commented:
> @tellthemachines thanks for your comments here, and they make totally sense but I'd like to handle the front-end in a different PR. Make sense for you?

The only front end issue I mentioned was the submenu styling one, the others are all affecting the editor experience. The main one is the previews for most of the default colour choices in the editor not working:

<img width="740" alt="Screen Shot 2019-11-05 at 10 49 42 am" src="https://user-images.githubusercontent.com/8096000/68167587-11414180-ffba-11e9-906a-af70c27e33ee.png">


I think this is happening because we are relying on classnames which have no CSS attached to them on the editor side, and we should be able to fix it by inlining the styles instead. That's what I was trying to say in my comment above, sorry if it wasn't clear.

retrofox commented:
> I think this is happening because we are relying on classnames which have no CSS attached to them on the editor side, and we should be able to fix it by inlining the styles instead. That's what I was trying to say in my comment above, sorry if it wasn't clear.

Thanks. Make sense. Sorry, you were; pretty clear. :-)

retrofox committed:
navigation-menu: always apply inline styles
In case the theme doesn't support CSS classes.

retrofox committed:
navigation-menu: fix lint errors

tellthemachines reviewed (approved):
This is working much better now! Approving the PR, and let's tackle the remaining issues in follow-up PRs.

Apart from the styling bugs I mentioned previously, there are a couple of accessibility problems we need to fix:
- Each colour palette should be wrapped in a `fieldset` with a `legend` instead of a plain `div` with a `span` for a label, so that screen readers can read out those labels.
- There's no way to close the dropdown once options are selected except for clicking outside or pressing Esc. Perhaps a close button would be useful?

obenland commented:
Thanks @tellthemachines!

@wunderbart Maybe one or both of those accessibility issues could be a good substitute for https://github.com/WordPress/gutenberg/issues/18135#issuecomment-550050257?

retrofox commented:
> Apart from the styling bugs I mentioned previously, there are a couple of accessibility problems we need to fix:
>
> Each colour palette should be wrapped in a fieldset with a legend instead of a plain div with a span for a label, so that screen readers can read out those labels.
> There's no way to close the dropdown once options are selected except for clicking outside or pressing Esc. Perhaps a close button would be useful?

Thank you very much for your help here. Let's create issues in order to avoid forgetting them.

mtias commented:
This change introduced a regression in how selected items displayed by forcing `font-size: $text-editor-font-size;`.
